esphome/components/ld2410/ld2410.cpp

```diff
@@ -95,40 +151,47 @@ void LD2410Component::handle_periodic_data_(uint8_t *buffer, int len) {
   if (buffer[7] != HEAD || buffer[len - 6] != END || buffer[len - 5] != CHECK)  // Check constant values
     return;  // data head=0xAA, data end=0x55, crc=0x00
 
-    /*
-      Data Type: 6th
-      0x01: Engineering mode
-      0x02: Normal mode
-    */
-    // char data_type = buffer[DATA_TYPES];
-    /*
-      Target states: 9th
-      0x00 = No target
-      0x01 = Moving targets
-      0x02 = Still targets
-      0x03 = Moving+Still targets
-    */
-#ifdef USE_BINARY_SENSOR
-  char target_state = buffer[TARGET_STATES];
-  if (this->target_binary_sensor_ != nullptr) {
-    this->target_binary_sensor_->publish_state(target_state != 0x00);
-  }
-#endif
-
   /*
     Reduce data update rate to prevent home assistant database size grow fast
   */
   int32_t current_millis = millis();
-  if (current_millis - last_periodic_millis < 1000)
+  if (current_millis - last_periodic_millis_ < this->throttle_)
     return;
-  last_periodic_millis = current_millis;
+  last_periodic_millis_ = current_millis;
 
+  /*
+    Data Type: 7th
+    0x01: Engineering mode
+    0x02: Normal mode
+  */
+  bool engineering_mode = buffer[DATA_TYPES] == 0x01;
+#ifdef USE_SWITCH
+  if (this->engineering_mode_switch_ != nullptr &&
+      current_millis - last_engineering_mode_change_millis_ > this->throttle_) {
+    if (engineering_mode && !this->engineering_mode_switch_->state) {
+      this->engineering_mode_switch_->turn_on();
+    } else if (!engineering_mode && this->engineering_mode_switch_->state) {
+      this->engineering_mode_switch_->turn_off();
+    }
+  }
```

Comment:
This should not call turn_on/off as its an external state change and should just be updating to match the hardware. It should just `publish_state`. `publish_state` already has a deduplicator so there is no risk calling again.

```suggestion
  if (this->engineering_mode_switch_ != nullptr) {
    this->engineering_mode_switch_->publish_state(engineering_mode);
  }
```

The same goes for the numbers and selects.

Reply:
select and numbers do not have dedup mechanism though, should I take that under account or not?

Reply:
done